### ISSUE TITLE
How to give the head_gradient properly when last layer is not loss layer?

### ISSUE THREAD

**Issue by yxchng** [apache/incubator-mxnet#3480]
``` python

def _bn_relu_conv(data, num_filter, kernel, stride = (1, 1), pad = (0, 0), act_type = 'relu'):
    bn = mx.symbol.BatchNorm(data)
    act = mx.symbol.Activation(bn, act_type = act_type)
    conv = mx.symbol.Convolution(act, num_filter = num_filter, kernel = kernel, stride = stride, pad = pad)
    return conv

def _conv_bn_relu(data, num_filter, kernel = (1, 1), stride = (1, 1), pad = (0, 0), act_type = 'relu'):
    conv = mx.symbol.Convolution(data, num_filter = num_filter, kernel = kernel, stride = stride, pad = pad)
    bn = mx.symbol.BatchNorm(conv)
    act = mx.symbol.Activation(bn, act_type = act_type)
    return act

def residual(data, num_filter, dim_match):
    conv1 = _bn_relu_conv(data, num_filter = num_filter / 2, kernel = (1, 1), pad = (1, 1))
    conv2 = _bn_relu_conv(conv1, num_filter = num_filter / 2, kernel = (3, 3))
    conv3 = _bn_relu_conv(conv2, num_filter = num_filter, kernel = (1, 1))
    if dim_match:
        return data + conv3
    else:
        skip = mx.symbol.Convolution(data, num_filter = num_filter, kernel = (1, 1), stride = (1, 1))
        return skip + conv3


def setup_model(ctx, model_symbol, batch_size, image_dim):
    model = {}
    model['symbol'] = model_symbol

    input_shape = (batch_size, image_dim[0], image_dim[1], image_dim[2])
    model_exec = model_symbol.simple_bind(ctx=ctx, data=input_shape, grad_req='add')
    model['executor'] = model_exec

    param_blocks = []
    arg_names = model_symbol.list_arguments()
    arg_dict = dict(zip(arg_names, model_exec.arg_arrays))
    args_grad = dict(zip(arg_names, model_exec.grad_arrays))
    for i, name in enumerate(arg_names):
        if name in ['data']:
            continue
        param_blocks.append((i, arg_dict[name], args_grad[name], name))
    model['param_blocks'] = param_blocks

    data = model_exec.arg_dict['data']
    model['data'] = data
    labels = []
    for i in model_exec.output_dict:
        labels.append(model_exec.output_dict[i])
    model['labels'] = labels

    model_symbol.save('model/model-symbol.json')

    return model


def train(model, batch_size, num_train_iteration, num_test_iteration, batch_iterator_train, batch_iterator_test, optimizer='rmsprop', learning_rate=2.5e-4, num_epoch=100, checkpoint_frequency=10):
    m = model


    # create optimizer
    opt = mx.optimizer.create(optimizer)
    opt.lr = learning_rate

    updater = mx.optimizer.get_updater(opt)

    for epoch in xrange(num_epoch):
        print "epoch", epoch
        tic = time.time()
        train_loss = np.array([0.0] * 9)

        num_train_images = 0
        for i in xrange(num_train_iteration):
            print i
            batch_data, batch_heatmaps = batch_iterator_train.next()

            m['data'][:] = batch_data
            m['labels'][:] = [batch_heatmaps]

            m['executor'].forward(is_train=True)
            forward_output = m['executor'].outputs
            m['executor'].backward(forward_output)

            for j in xrange(8):
                train_loss[j] += ((forward_output[j].asnumpy() - batch_heatmaps) ** 2.0).mean()
            train_loss[8] = sum(train_loss[0:8])

            print train_loss
            norm = 0

            # update weights
            for idx, weight, grad, name in m['param_blocks']:
                grad /= batch_size
                updater(idx, grad, weight)
                grad[:] = 0.0

        toc = time.time()
        train_time = toc - tic

        train_loss_average = train_loss / float(num_train_iteration)

        if (epoch + 1) % checkpoint_frequency == 0:
            save_dict = {('arg:%s' % k): v for k, v in m['executor'].arg_dict.items()}
            save_dict.update({('aux:%s' % k): v for k, v in m['executor'].aux_dict.items()})
            param_name = 'checkpoint/model-%04d.params' % (prefix, iteration)
            mx.nd.save(param_name, save_dict)
            print >> logs, 'Saved checkpoint to %s' % param_name

        test_loss = np.array([0.0] * 9)
        num_test_images = 0    

        for i in xrange(num_test_iteration):
            batch_data, batch_heatmaps = batch_iterator_test.next()

            m['data'][:] = batch_data
            m['executor'].forward(is_train=False)
            forward_output = m['executor'].outputs

            for j in xrange(8):
                test_loss[j] += ((forward_output[j].asnumpy() - batch_heatmaps) ** 2.0).mean()
            test_loss[8] = sum(test_loss[0:8])
        test_loss_average = test_loss / float(num_test_iteration)

        print train_loss_average, test_loss_average


model_symbol = create_model_symbol(8, 1)
# model_symbol = get_symbol()

stacked_hourglass = setup_model(mx.gpu(), model_symbol, 6, (3, 256, 256))

batch_iterator_train = BatchIterator('train_fuji.h5', 6, 8)
batch_iterator_test = BatchIterator('test_fuji.h5', 6, 8)
train(stacked_hourglass, 6, 50, 50, batch_iterator_train, batch_iterator_test)
```

Above is my code. I am not sure if my head gradient is provided properly. Even if it is not, right now, the training is not working. For all the 8 outputs, the predicted values are the same. This is definitely not right. Does anyone know how can I train a network whereby the last layer is not a loss layer?

**Comment by yxchng** [comment]
#3460

**Comment by winstywang** [comment]
You can simply pass the head gradient in m['executor'].backward(grad) where the grad is the head gradient. It should has the same shape as the outputs of the executor.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
